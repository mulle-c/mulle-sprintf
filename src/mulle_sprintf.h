/*
 *  MulleFoundation - A tiny Foundation replacement
 *
 *  mulle_sprintf.m is a part of MulleFoundation
 *
 *  Copyright (C) 2011 Nat!, Mulle kybernetiK.
 *  All rights reserved.
 *
 *  Coded by Nat!
 *
 *  $Id$
 *
 */
#ifndef mulle_sprintf_h__
#define mulle_sprintf_h__

#include "mulle_sprintf_function.h"
#include <mulle_vararg/mulle_vararg.h>


#define MULLE_SPRINTF_VERSION  ((0 << 20) | (6 << 8) | 0)

struct mulle_buffer;
struct mulle_sprintf_conversion;


int   mulle_sprintf( struct mulle_buffer *buffer,
                     char *format,
                     ...);

int   mulle_vsprintf( struct mulle_buffer *buffer,
                      char *format,
                      va_list va);

int   _mulle_vsprintf( struct mulle_buffer *buffer,
                       char *format,
                       va_list va,
                       struct mulle_sprintf_conversion *table);

int   mulle_mvsprintf( struct mulle_buffer *buffer,
                       char *format,
                       mulle_vararg_list va);

int   _mulle_mvsprintf( struct mulle_buffer *buffer,
                        char *format,
                        mulle_vararg_list arguments,
                        struct mulle_sprintf_conversion *table);

#pragma mark -
#pragma manage "static" storage

extern void   *(*mulle_sprintf_get_storage)( struct mulle_allocator *);
extern void   (*mulle_sprintf_free_storage)( void);


#if MULLE_ALLOCATOR_VERSION < ((1 << 20) | (5 << 8) | 0)
# error "mulle_allocator is too old"
#endif
#if MULLE_CONTAINER_VERSION < ((0 << 20) | (5 << 8) | 0)
# error "mulle_container is too old"
#endif

#endif
