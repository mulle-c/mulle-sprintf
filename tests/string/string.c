#include <mulle_sprintf/mulle_sprintf.h>
#include <mulle_test_allocator/mulle_test_allocator.h>
#include <mulle_container/mulle_container.h>

#include <stdio.h>



static void   simple_string_test()
{
   struct mulle_buffer   *buffer;

   buffer = mulle_buffer_create( NULL);


   mulle_sprintf( buffer, "%s\n", "VfL Bochum 1848");
   printf( "%s", mulle_buffer_get_bytes( buffer));
   mulle_buffer_reset( buffer);

   mulle_sprintf( buffer, "%.3s\n", "VfL Bochum 1848");
   printf( "%s", mulle_buffer_get_bytes( buffer));
   mulle_buffer_reset( buffer);

   mulle_sprintf( buffer, "\"%.100s\"\n", "VfL Bochum 1848");
   printf( "%s", mulle_buffer_get_bytes( buffer));
   mulle_buffer_reset( buffer);

   mulle_sprintf( buffer, "\"%100s\"\n", "VfL Bochum 1848");
   printf( "%s", mulle_buffer_get_bytes( buffer));
   mulle_buffer_reset( buffer);

   mulle_sprintf( buffer, "\"%100.100s\"\n", "VfL Bochum 1848");
   printf( "%s", mulle_buffer_get_bytes( buffer));
   mulle_buffer_reset( buffer);

   mulle_sprintf( buffer, "\"%.*s\"\n", 100, "VfL Bochum 1848");
   printf( "%s", mulle_buffer_get_bytes( buffer));
   mulle_buffer_reset( buffer);

   mulle_sprintf( buffer, "\"%*s\"\n", 100, "VfL Bochum 1848");
   printf( "%s", mulle_buffer_get_bytes( buffer));
   mulle_buffer_reset( buffer);

   mulle_sprintf( buffer, "\"%*.*s\"\n", 100, 100, "VfL Bochum 1848");
   printf( "%s", mulle_buffer_get_bytes( buffer));
   mulle_buffer_reset( buffer);

   mulle_sprintf( buffer, "%s", "");
   printf( "%s", mulle_buffer_get_bytes( buffer));
   mulle_buffer_reset( buffer);

   mulle_buffer_destroy( buffer);
}


int  main()
{
   mulle_test_allocator_set_tracelevel( 1);
   mulle_test_allocator_initialize();
   mulle_default_allocator = mulle_test_allocator;

   simple_string_test();

   mulle_sprintf_free_storage();
   mulle_test_allocator_reset();
}

