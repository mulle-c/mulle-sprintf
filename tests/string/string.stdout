m: |VfL Bochum 1848|
m: |VfL|
m: |"VfL Bochum 1848"|
m: |"                                                                                     VfL Bochum 1848"|
m: |"                                                                                     VfL Bochum 1848"|
m: |"VfL Bochum 1848"|
m: |"                                                                                     VfL Bochum 1848"|
m: |"                                                                                     VfL Bochum 1848"|
m: ||
p: |VfL Bochum 1848|
p: |VfL|
p: |"VfL Bochum 1848"|
p: |"                                                                                     VfL Bochum 1848"|
p: |"                                                                                     VfL Bochum 1848"|
p: |"VfL Bochum 1848"|
p: |"                                                                                     VfL Bochum 1848"|
p: |"                                                                                     VfL Bochum 1848"|
p: ||
